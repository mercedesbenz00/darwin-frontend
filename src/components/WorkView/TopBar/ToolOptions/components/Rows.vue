<template>
  <input
    v-model="_value"
    v-input-auto-blur="true"
    class="rows"
    @keydown.stop
    @keypress.stop="onlyNumber"
    @keyup.stop
  >
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import { Editor } from '@/engine/editor'

@Component({ name: 'rows' })
export default class Rows extends Vue {
  @Prop({ required: true })
  editor!: Editor

  @Prop({ required: true, type: Number })
  value!: number

  get _value (): string { return this.value.toString() }
  set _value (value: string) {
    const size = parseInt(value, 10)
    if (size <= 0 || isNaN(size)) {
      this.$store.dispatch('toast/notify', {
        content: 'Number of rows should be a number greater than 0'
      })
      return
    }
    this.editor.callCommand('table_tool.set_rows', size)
  }

  onlyNumber (event: KeyboardEvent): void {
    const keyCode = (event.keyCode ? event.keyCode : event.which)
    if ((keyCode < 48 || keyCode > 57) && keyCode !== 46 && keyCode !== 13 && keyCode !== 8) {
      event.preventDefault()
    }
  }
}
</script>

<style lang="scss" scoped>
.rows {
  @include row--center;
  width: 100%;
  height: 100%;
  padding: 0 10px;
  background: $colorSecondaryLight2;
  border-radius: 3px 0 0 3px;
  text-align: center;
  user-select: none;
}
</style>
