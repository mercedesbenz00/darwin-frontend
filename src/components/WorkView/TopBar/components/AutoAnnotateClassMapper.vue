<template>
  <div class="auto-annotate-class-mapper">
    <secondary-button
      class="auto-annotate-class-mapper__button"
      size="medium"
      @click="openMappingDialog"
    >
      Map Classes
    </secondary-button>
    <modal
      name="auto-annotate-mapping"
      height="auto"
      :width="500"
    >
      <class-mapping
        :annotation-classes="datasetClasses"
        :running-session="session"
        @cancel="closeMappingDialog"
        @confirm="updateMapping"
      />
    </modal>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import ClassMapping from '@/components/DatasetSettings/ModelStage/ClassMapping.vue'
import { MappedClass } from '@/components/DatasetSettings/ModelStage/types'
import { toClassMapping } from '@/components/DatasetSettings/ModelStage/utils'
import { Editor } from '@/engine/editor'
import { updateClassMapping } from '@/store/modules/workview/actions/updateClassMapping'
import {
  AnnotationClassPayload,
  DatasetPayload,
  RootState,
  RunningSessionPayload,
  StoreActionPayload
} from '@/store/types'

@Component({ name: 'auto-annotate-class-mapper', components: { ClassMapping } })
export default class AutoAnnotateClassMapper extends Vue {
  @Prop({ required: true, type: Object as () => Editor })
  editor!: Editor

  @State((state: RootState) => state.aclass.classes)
  allClasses!: AnnotationClassPayload[]

  @State(state => state.workview.dataset)
  dataset!: DatasetPayload

  get datasetClasses (): AnnotationClassPayload[] {
    const { allClasses, dataset } = this
    return allClasses.filter(c => c.datasets.some(d => d.id === dataset.id))
  }

  get session (): RunningSessionPayload | null {
    return this.editor.preselectedAutoAnnotateModel
  }

  openMappingDialog (): void {
    this.$modal.show('auto-annotate-mapping')
  }

  closeMappingDialog (): void {
    this.$modal.hide('auto-annotate-mapping')
  }

  updateMapping (mapping: MappedClass[]): void {
    const { session } = this
    if (!session) { return }

    const classMapping = mapping.map(toClassMapping)
    const payload: StoreActionPayload<typeof updateClassMapping> = {
      classMapping,
      runningSessionId: session.id
    }
    this.$store.dispatch('workview/updateClassMapping', payload)

    this.closeMappingDialog()
  }
}
</script>

<style lang="scss" scoped>
.auto-annotate-class-mapper {
  margin: 0 10px;
}

.auto-annotate-class-mapper__button {
  @include typography(md, defult, 500);
  font-family: $fontFamilyInter;
  letter-spacing: 0;
  width: 100px;
  height: 32px;
  padding: 8px;
  border: 1px solid $colorInteractiveSecondaryHover;
  box-sizing: border-box;
  border-radius: 8px;
}

.auto-annotate-class-mapper__button:hover {
  background-color: $colorOverlayHover;
  border: 1.5px solid $colorInteractiveSecondaryHover;
}
</style>
