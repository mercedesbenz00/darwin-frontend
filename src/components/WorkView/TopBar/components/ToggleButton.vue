<template>
  <v-popover
    :delay="100"
    placement="bottom"
    popover-class="toggle-button-popover"
    trigger="hover"
  >
    <button
      class="toggle-button"
      @click="$emit('toggle')"
    >
      <slot />
    </button>

    <template slot="popover">
      <slot name="tooltip" />
    </template>
  </v-popover>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({ name: 'top-bar-toggle-button' })
export default class TopBarToggleButton extends Vue {
  @Prop({ required: true, type: Boolean })
  on!: boolean
}
</script>

<style lang="scss" scoped>
.toggle-button {
  @include row--center;
  background: transparent;
  position: relative;
  padding: 4px 5px;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  transition: background .2s ease;

  &:hover {
    background: $colorAliceShade;
  }

  &:active {
    background: $colorAliceShadow;
  }
}
</style>
