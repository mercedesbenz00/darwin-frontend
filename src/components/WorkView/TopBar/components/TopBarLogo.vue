<template>
  <div class="top-bar__logo">
    <icon-button
      v-tooltip="tooltip"
      class="top-bar__logo__button"
      size="small"
      tag="router-link"
      exact
      :to="backUrl"
    >
      <img
        v-if="teamLogo"
        :src="teamLogo"
        class="top-bar__logo__button__img top-bar__logo__button__img--normal"
      >
      <logo-icon
        v-else
        class="top-bar__logo__button__img top-bar__logo__button__img--logo"
      />
      <back-icon
        class="top-bar__logo__button__img top-bar__logo__button__img--hover"
      />
    </icon-button>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { Location } from 'vue-router'
import { State } from 'vuex-class'

import { IconButton } from '@/components/Common/Button/V2'
import { LogoIcon, BackIcon } from '@/components/WorkView/TopBar/assets/icons'
import { WorkviewState } from '@/store/modules/workview/state'
import { RootState, TeamPayload } from '@/store/types'
import { TooltipOptions } from '@/types'
import { resolveWorkviewLocation } from '@/utils'

@Component({
  name: 'top-bar-logo',
  components: { IconButton, LogoIcon, BackIcon }
})
export default class TopBarLogo extends Vue {
  @State(state => state.team.currentTeam)
  currentTeam!: TeamPayload

  @State((state: RootState) => state.workview.dataset)
  dataset!: WorkviewState['dataset']

  @State((state: RootState) => state.workview.tutorialMode)
  tutorialMode!: boolean

  get tooltip (): TooltipOptions {
    return {
      content: '<b>Back</b>',
      placement: 'right',
      delay: { show: 300 }
    }
  }

  get teamLogo (): string {
    if (!this.currentTeam || !this.currentTeam.image) { return '' }
    return this.currentTeam.image.thumbnail_url || this.currentTeam.image.url
  }

  get backUrl (): Location {
    const { dataset, tutorialMode } = this
    if (!dataset) { return { path: '/' } }
    if (tutorialMode) { return { path: '/datasets' } }

    return resolveWorkviewLocation(this.$route, dataset)
  }
}
</script>

<style lang="scss" scoped>
$logoWidth: 44px;

.top-bar__logo {
  @include col--center;
  width: $logoWidth;
  height: $logoWidth;

  &__button {
    @include row--center;
    background-color: $colorSurfaceInverted;
    background-color: red;
    padding: 0;
    max-width: $logoWidth;
    max-height: $logoWidth;
    overflow: hidden;
    border-radius: 50%;
    cursor: pointer;

    &:hover:not(:disabled),
    &:active:not(:disabled) {
      .top-bar__logo__button__img {
        &--normal,
        &--logo {
          display: none;
        }

        &--hover {
          display: block;
        }
      }
    }

    .top-bar__logo__button__img {
      &--normal,
      &--logo {
        display: block;
      }

      &--logo,
      &--hover {
        border-radius: 100%;
      }

      &--normal {
        display: block;
        width: 32px;
        height: 32px;
        object-fit: cover;
        border-radius: 8px;
      }

      &--hover {
        display: none;
      }
    }
  }
}
</style>
