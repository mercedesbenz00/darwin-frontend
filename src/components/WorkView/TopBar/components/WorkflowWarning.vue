<template>
  <div class="work-tracker">
    <warning-icon
      v-if="channelStatus === 'error'"
      v-tooltip="tooltip"
      class="work-tracker__warning"
    />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Watch } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { WarningIcon } from '@/components/WorkView/TopBar/assets/icons'
import { WorkviewTrackerState } from '@/store/modules/workviewTracker'
import { TooltipOptions } from '@/types'

@Component({ name: 'workflow-warning', components: { WarningIcon } })
export default class Warning extends Vue {
  @Prop({ required: false, default: null, type: String as () => string })
  topic!: string

  @State(state => state.workviewTracker.status)
  channelStatus!: WorkviewTrackerState['status']

  @Watch('channelStatus')
  onChannelStatus (
    status: WorkviewTrackerState['status'] | null,
    previous: WorkviewTrackerState['status'] | null
  ): void {
    if (status !== 'error') { return }

    this.$store.dispatch('toast/warning', {
      content: (
        "We couldn't initialize work tracking. " +
        'Your time will not be tracked. Try refreshing your browser.'
      )
    })

    this.$sentry.captureEvent({
      message: "Tracker couldn't connect to workview channel",
      extra: {
        topic: this.topic,
        oldChannelStatus: previous,
        newChannelStatus: status,
        trackerState: this.$store.state.workviewTracker
      }
    })
  }

  get tooltip (): TooltipOptions {
    return {
      content: `
        <p>There was an issue connecting to work tracking. Yout time will not be tracked.</p>
        <p>Try refreshing your browser, or switching to a different browser.</p>
        <p>If the problem persists, contact support.</p>
      `,
      classes: 'tooltip--work-tracker',
      placement: 'bottom',
      delay: { show: 300, hide: 300 }
    }
  }
}
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.tooltip--work-tracker {
  .tooltip-inner {
    display: table;
  }
}
</style>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.fade-enter-active, .fade-leave-active {
  transition: opacity .5s;
}
.fade-enter, .fade-leave-to {
  opacity: 0;
}
</style>

<style lang="scss" scoped>
.work-tracker {
  @include row--distributed;

  &__warning {
    @include row--center;
    justify-self: flex-start;

    &:focus,
    &:active {
      outline: none;
    }
  }
}
</style>
