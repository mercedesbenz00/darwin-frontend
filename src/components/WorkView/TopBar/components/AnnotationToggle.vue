<template>
  <toggle-button
    class="top-bar__toggle-annotations"
    :class="{ 'top-bar__toggle-annotations--on': on }"
    :on="on"
    @toggle="toggle"
  >
    <icon-mono-hide v-if="on" />
    <icon-mono-eye v-else />

    <template #tooltip>
      <b class="top-bar__toggle-annotations__tooltip">
        Toggle Annotations <span class="tooltip__hotkey">H</span>
      </b>
    </template>
  </toggle-button>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { IconMonoEye, IconMonoHide } from '@/assets/icons/V2/Mono'

import ToggleButton from './ToggleButton.vue'

@Component({
  name: 'annotation-toggle',
  components: { IconMonoEye, IconMonoHide, ToggleButton }
})
export default class AnnotationToggle extends Vue {
  @State(state => state.workview.renderAnnotations)
  on!: boolean

  toggle (): void {
    this.$store.commit('workview/TOGGLE_ANNOTATIONS')
  }
}
</script>

<style lang="scss" scoped>
.top-bar__toggle-annotations {
  :deep(svg) {
    pointer-events: none;
    width: 20px;

    path {
      fill: $colorContentTertiary;
      stroke: $colorContentTertiary;
    }
  }
}

.top-bar__toggle-annotations__tooltip {
  @include row;
  align-items: center;
}
</style>
