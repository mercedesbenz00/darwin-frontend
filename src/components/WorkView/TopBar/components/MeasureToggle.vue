<template>
  <toggle-button
    class="top-bar__toggle-measures"
    :class="{ 'top-bar__toggle-measures--on': on }"
    :on="on"
    @toggle="toggle"
  >
    <icon-mono-measurements />

    <template #tooltip>
      <b class="top-bar__toggle-measures__tooltip">
        Toggle Measurements <span class="tooltip__hotkey">{{ hotkeyInfo }}</span>
      </b>
    </template>
  </toggle-button>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { IconMonoMeasurements } from '@/assets/icons/V2/Mono'
import { RootState } from '@/store/types'
import { onMacOS } from '@/utils'

import ToggleButton from './ToggleButton.vue'

@Component({
  name: 'measures-toggle',
  components: { IconMonoMeasurements, ToggleButton }
})
export default class MeasuresToggle extends Vue {
  @State((state: RootState) => state.workview.renderMeasures)
  on!: boolean

  get hotkeyInfo (): string {
    return onMacOS() ? 'âŒ˜ + M' : 'CTRL + M'
  }

  toggle (): void {
    this.$store.commit('workview/TOGGLE_MEASURES')
  }
}
</script>

<style lang="scss" scoped>
.top-bar__toggle-measures {
  :deep(svg) {
    pointer-events: none;
    width: 20px;
    path {
      fill: $colorContentTertiary;
      stroke: $colorContentTertiary;
    }
  }
}

.top-bar__toggle-measures__tooltip {
  @include row;
  align-items: center;
}
</style>
