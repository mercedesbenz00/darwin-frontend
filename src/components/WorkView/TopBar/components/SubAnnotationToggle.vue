<template>
  <toggle-button
    class="top-bar__toggle-sub-annotations"
    :class="{ 'top-bar__toggle-sub-annotations--on': on }"
    :on="on"
    @toggle="toggle"
  >
    <icon-mono-centroid />

    <template #tooltip>
      <b class="top-bar__toggle-sub-annotations__tooltip">
        Toggle Sub Annotations <span class="tooltip__hotkey">SHIFT+H</span>
      </b>
    </template>
  </toggle-button>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { IconMonoCentroid } from '@/assets/icons/V2/Mono'

import ToggleButton from './ToggleButton.vue'

@Component({
  name: 'sub-annotation-toggle',
  components: { IconMonoCentroid, ToggleButton }
})
export default class SubAnnotationToggle extends Vue {
  @State(state => state.workview.renderSubAnnotations)
  on!: boolean

  toggle (): void {
    this.$store.commit('workview/TOGGLE_SUBANNOTATIONS')
  }
}
</script>

<style lang="scss" scoped>
.top-bar__toggle-sub-annotations {
  :deep(svg) {
    pointer-events: none;
    height: 15px;

    rect {
      stroke: $colorContentTertiary;
    }
  }
}

.top-bar__toggle-sub-annotations__tooltip {
  @include row;
  align-items: center;
}
</style>
