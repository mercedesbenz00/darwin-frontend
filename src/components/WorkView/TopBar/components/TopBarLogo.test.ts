import { createLocalVue, shallowMount, Stubs } from '@vue/test-utils'
import Vuex from 'vuex'

import { createMockTheme } from 'test/unit/components/mocks'
import createTestStore, { giveFullMemberAbility, giveAnnotatorAbiliy } from 'test/unit/createTestStore'
import { buildTeamPayload, buildDatasetPayload } from 'test/unit/factories'

import TopBarLogo from '@/components/WorkView/TopBar/components/TopBarLogo.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('tooltip', () => {})

let store: ReturnType<typeof createTestStore>
const stubs: Stubs = { 'router-link': true }
let mocks: {
  $router: { go: () => void }
  $route: { path: string, query: Object }
  $theme: ReturnType<typeof createMockTheme>
}
let propsData: { disabled?: boolean }

const sfh = buildDatasetPayload({ id: 1, slug: 'SFH', team_slug: 'V7' })

const openDatasetWorkviewRoute: (typeof mocks)['$route'] = { path: '/V7/SFH/1', query: {} }
const workviewRoute: (typeof mocks)['$route'] = { path: '/workview', query: {} }

beforeEach(() => {
  store = createTestStore()
  store.commit('workview/SET_DATASET', sfh)
  giveFullMemberAbility(store)

  mocks = {
    $route: workviewRoute,
    $router: { go: () => {} },
    $theme: createMockTheme()
  }
  propsData = {}
})

const v7 = buildTeamPayload({
  name: 'V7', image: { id: 1, thumbnail_url: 'v7_thumb.png', url: 'v7.png' }
})

it('matches snapshot', () => {
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with team logo available', () => {
  store.commit('team/SET_CURRENT_TEAM', v7)
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot on workview', () => {
  mocks.$route = workviewRoute
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot on open datasets', () => {
  mocks.$route = openDatasetWorkviewRoute
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when annotator', () => {
  giveAnnotatorAbiliy(store)
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot in tutorial', () => {
  store.commit('workview/SET_TUTORIAL_MODE', true)
  const wrapper = shallowMount(TopBarLogo, { localVue, mocks, propsData, store, stubs })
  expect(wrapper).toMatchSnapshot()
})
