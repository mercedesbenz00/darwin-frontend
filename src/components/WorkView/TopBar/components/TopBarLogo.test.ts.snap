// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button">
    <logo-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--logo"></logo-icon-stub>
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;

exports[`matches snapshot in tutorial 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button">
    <logo-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--logo"></logo-icon-stub>
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;

exports[`matches snapshot on open datasets 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button">
    <logo-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--logo"></logo-icon-stub>
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;

exports[`matches snapshot on workview 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button">
    <logo-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--logo"></logo-icon-stub>
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;

exports[`matches snapshot when annotator 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button">
    <logo-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--logo"></logo-icon-stub>
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;

exports[`matches snapshot with team logo available 1`] = `
<div class="top-bar__logo">
  <icon-button-stub variant="default" color="secondary" size="small" flair="soft" tag="router-link" exact="" to="[object Object]" class="top-bar__logo__button"><img src="v7_thumb.png" class="top-bar__logo__button__img top-bar__logo__button__img--normal">
    <back-icon-stub class="top-bar__logo__button__img top-bar__logo__button__img--hover"></back-icon-stub>
  </icon-button-stub>
</div>
`;
